javascript/jose/security/jwt-exposed-credentials.yaml

```diff
@@ -4,230 +4,52 @@ rules:
-          var { JWT } = $JOSE;
-          ...
-          $P.$U.password = ...;
-          ...
-          var $T = JWT.sign($P,...);
-      - pattern: |
-          var $JOSE = require("jose");
-          ...
-          var { JWT } = $JOSE;
-          ...
-          $P.$U.password = ...;
-          ...
-          $T = JWT.sign($P,...);
-      - pattern: |
-          var $JOSE = require("jose");
-          ...
-          var { JWT } = $JOSE;
-          ...
-          var $P = Object.assign(...,{$U:{password:...}},...);
-          ...
-          var $T = JWT.sign($P,...);
-      - pattern: |
-          var $JOSE = require("jose");
-          ...
-          var { JWT } = $JOSE;
-          ...
-          var $P = Object.assign(...,{$U:{password:...}},...);
-          ...
-          $T = JWT.sign($P,...);
-      - pattern: |
-          var $JOSE = require("jose");
-          ...
-          var { JWT } = $JOSE;
-          ...
-          $P = Object.assign(...,{$U:{password:...}},...);
-          ...
-          var $T = JWT.sign($P,...);
-      - pattern: |
-          var $JOSE = require("jose");
-          ...
-          var { JWT } = $JOSE;
-          ...
-          $P = Object.assign(...,{$U:{password:...}},...);
-          ...
-          $T = JWT.sign($P,...);
-      - pattern: |
-          var $JOSE = require("jose");
-          ...
-          var { JWT } = $JOSE;
-          ...
-          var $T = JWT.sign(Object.assign(...,{$U:{password:...}},...),...);
-      - pattern: |
-          var $JOSE = require("jose");
-          ...
-          var { JWT } = $JOSE;
-          ...
-          $T = JWT.sign(Object.assign(...,{$U:{password:...}},...),...);
+    message: >- 
+      A hard-coded credential was detected. It is not recommended 
+      to store credentials in source-code, as this risks secrets
+      being leaked and used by either an internal or external malicious adversary. 
+      It is recommended to use environment variables to 
+      securely provide credentials or retrieve credentials from 
+      a secure vault or HSM (Hardware Security Module).
+    severity: WARNING
+    mode: taint
+    pattern-sources:
+        - patterns:
+            - pattern-either:
+                - pattern: |
+                    {...,password: '...',...}
+                - pattern: $OBJ. ... .password = '...'
+    pattern-sinks:
+        - patterns:
+            - pattern-either:
+                - pattern-inside: |
+                    $JWT = require("jsonwebtoken")
+                    ...
+                - pattern-inside: |
+                    import $JWT from "jsonwebtoken"
+                    ...
+                - pattern-inside: |
+                    import * as $JWT from "jsonwebtoken"
+                    ...
+                - pattern-inside: |
+                    import {...,$JWT,...} from "jsonwebtoken"
+                    ...
+            - pattern-either:
+                - pattern-inside: |
+                    $JWT.sign($F,...);
```

Comment:
the signing key for jsonwebtoken is the second positional arg: https://github.com/auth0/node-jsonwebtoken/blob/master/sign.js#L82

This rule seems to be for accidentally adding passwords to the signed payload (name is `jwt-exposed-credentials`), but the messaging is for secrets in source.

The case that this finds is *highly* unusual and I think it ought to be two rules:

* rule for finding "password" in JWT payload (info disclosure)
* rule for hardcoded signing key